polyply/tests/test_nb_engine.py

```diff
@@ -0,0 +1,208 @@
+    test system
+    [ molecules ]
+    testA 2
+    testB 1
+    """
+    lines = textwrap.dedent(top_lines)
+    lines = lines.splitlines()
+    force_field = ForceField("test")
+    topology = Topology(force_field)
+    read_topology(lines=lines, topology=topology, cwdir="./")
+    topology.preprocess()
+    topology.volumes = {"GLY": 0.53, "GLU": 0.67, "ASP": 0.43}
+    return topology
+
+@pytest.mark.parametrize('n_pos', (
+      0,
+      3,
+      5))
+def  test_nb_engine_from_top_file(topology, n_pos):
+
+     # we add positions to the topology skipping some
+     # positins depending on how many are set in n_pos
+     positions = np.random.random(n_pos*3).reshape(-1, 3)
+     for mol_idx, mol in enumerate(topology.molecules):
+         for node in mol.nodes:
+             try:
+                mol.nodes[node]["position"] = positions[mol_idx + node]
+             except IndexError:
+                continue
+
+     # initiate the nb_engine
+     nb_engine =  NonBondEngine.from_topology(topology.molecules,
+                                              topology,
+                                              box=np.array([10., 10., 10.]))
+
+     # these are the expected interactions in the nb_engine
+     nb_interactions = {
+                        frozenset(["GLY", "GLY"]): 0.53,
+                        frozenset(["GLY", "GLU"]): (0.53+0.67)/2.0,
+                        frozenset(["GLY", "ASP"]): (0.43+0.53)/2.0,
+                        frozenset(["ASP", "ASP"]): 0.43,
+                        frozenset(["ASP", "GLU"]): (0.43+0.67)/2.0,
+                        frozenset(["GLU", "GLU"]): 0.67,
+                       }
+
+     # check if all interactions are created properly
+     for res_combo, interaction in nb_interactions.items():
+          assert nb_engine.interaction_matrix[res_combo] == (nb_interactions[res_combo], 1.0)
+
+     # make sure that all positions are in the global position matrix
+     # and positions which are undefined have a infinity set
+     for mol_idx, mol in enumerate(topology.molecules):
+         for node in mol.nodes:
+            if "position" in mol.nodes[node]:
+                assert all(nb_engine.get_point(mol_idx, node) == mol.nodes[node]["position"])
+            else:
+                assert all(nb_engine.positions[mol_idx, node] == np.array([np.inf, np.inf, np.inf]))
+
+def test_add_positions(topology):
+     # we add 2 positions to the nb_matrix
+     positions = np.random.random(2*3).reshape(-1, 3)
+     for mol_idx, mol in enumerate(topology.molecules):
+         for node in mol.nodes:
+             try:
+                mol.nodes[node]["position"] = positions[mol_idx + node]
+             except IndexError:
+                continue
+
+     nb_engine =  NonBondEngine.from_topology(topology.molecules,
+                                              topology,
+                                              box=np.array([10., 10., 10.]))
+
+     # now we check if the 8th is added correctly
```

Comment:
3rd?

Reply:
yes I was probably thinking of the atomsitic molecule underlying this one.